File: PaperLib/Repositories/Scrapers/Scraper.swift
//
//  Scraper.swift
//  PaperLib
//
//  Created by GeoffreyChen on 15/02/2022.
//

import Foundation
import Alamofire
import Combine

protocol Scraper {
    func preProcess(entityDraft: PaperEntityDraft) -> (String, HTTPHeaders, Bool)
    func parsingProcess(response: String?, entityDraft: PaperEntityDraft) -> AnyPublisher<PaperEntityDraft, WebError>
    func scrape(entityDraft: PaperEntityDraft) -> AnyPublisher<PaperEntityDraft, WebError>
    func scrapeImpl(entityDraft: PaperEntityDraft) -> AnyPublisher<PaperEntityDraft, WebError>
}

extension Scraper {
    func scrape(entityDraft: PaperEntityDraft) -> AnyPublisher<PaperEntityDraft, WebError> {
        return scrapeImpl(entityDraft: entityDraft)
    }

    func scrapeImpl(entityDraft: PaperEntityDraft) -> AnyPublisher<PaperEntityDraft, WebError> {
        let (scrapeURL, headers, enable) = preProcess(entityDraft: entityDraft)
        if enable {
            return AF.request(scrapeURL, headers: headers)
                .publishString()
                .flatMap {
                    parsingProcess(response: $0.value, entityDraft: entityDraft)
                }
                .eraseToAnyPublisher()
        } else {
            return Just<PaperEntityDraft>
                .withErrorType(entityDraft, WebError.self)
                .eraseToAnyPublisher()
        }
    }
}

File: src/repositories/scrapers/scraper.js
import got from 'got';


export class Scraper {
  async scrape(entityDraft) {
    try {
      entityDraft = await this.scrapeImpl(entityDraft);
    } catch (error) {
      this.sharedState.set('viewState.alertInformation', `Scraper Error: ${error}`);
    }
    return entityDraft;
  }

  preProcess(entityDraft) {
    throw new Error('Method not implemented.');
  }

  parsingProcess(rawResponse, entityDraft) {
    throw new Error('Method not implemented.');
  }

  async scrapeImpl(entityDraft) {
    const {scrapeURL, headers, enable} = this.preProcess(entityDraft);
    if (enable) {
      const response = await got(
          scrapeURL,
          {
            headers: headers,
          },
      );
      return this.parsingProcess(response, entityDraft);
    } else {
      return entityDraft;
    }
  }
}
